Update routes and affected components in frontend

amandahamynen opened the pull request:
## Created three new components:
### frontend/src/components/ProtectedRoute.jsx
  - Used for routes that require authentication, eliminates repetitive code checks (```{isLoggedIn ? … ```).
  - Redirects unauthenticated users to _/login_, otherwise renders the child route with ```<Outlet />```

### frontend/src/components/NavigationBar.jsx
  - Moved the navigation bar into its own component to clean up App.jsx.
  - Uses props ```isLoggedIn``` and ```handleLogout```

### frontend/src/pages/GrowerSitesPage.jsx
  - Created from previous ```SitePage.jsx```
  - Updated to use correct params ```params.siteId``` 
  - Updated the original function ```deleteSite``` to navigate to correct url after site deletion since ```GrowerSitesPage.jsx``` is used in two routes; _/grower/sites_ and _/grower/:siteId/sites_

## Updated the following:

### frontend/src/App.jsx
  - Routes that require authentication now make use of ProtectedRoutes-component 
  ```js
  <Route element={<ProtectedRoute isLoggedIn={isLoggedIn} />}>

      {/* Routes that require authentication go here */}

  </Route>
  ```
  - The navigation bar was refactored into its own component 
  ```js
  <NavigationBar isLoggedIn={isLoggedIn} handleLogout={handleLogout}/>
  ```
  - Refactored setting language and getting default role into their own functions ```setLanguage``` and ```getDefaultRole```
  - New routes are as follows:
      - _/grower_
          - Renders ```<GrowerHomePage />```
          - Empty except for the title ”Home”
      - _/grower/flowers_
          - Renders ```<GrowerFlowerPage />```
          - Lists all of the user’s own flowers using ```GrowerFlowerList.jsx```
      - _/grower/sites_
          - Renders ```<GrowerSitesPage />```
          - Lists all root sites, and the grower is able to add new sites
      - _/grower/:siteId_
          - Renders ```<GrowerHomePage />```
          - Empty except for the name and id of the site whose id is given as parameter ```params.siteId```
      - _/grower/:siteId/flowers_
          - Renders ```<GrowerFlowerPage />```
          - Lists all flowers related to the site whose id is given as parameter ```params.siteId```. These flowers include the site’s flowers and all of its subsites’ flowers.
      - _/grower/:siteId/sites_
          - Renders ```<GrowerSitesPage />```
          - Shows the name of the site whose id is given as parameter ```params.siteId```. Grower can add flowers and new subsites to that site, and also delete the site.

### frontend/src/layouts/GrowerLayout.jsx
  - Added new NavLink ”Sites” to the side navigation bar
      - if ```GrowerLayout.jsx``` is used on pages that are accessed through urls that start with _/grower/:siteId_,
          - ”Home” redirects to _/grower/:siteId_
          - ”Flowers” redirects to _/grower/:siteId/flowers_
          - ”Sites” redirects to _/grower/:siteId/sites_
      - If ```GrowerLayout.jsx``` is used on pages that are accessed through urls that start with _/grower_ (without params),
          - ”Home” redirects to _/grower_
          - ”Flowers” redirects to _/grower/flowers_
          - ”Sites” redirects to _/grower/sites_

### frontend/src/components/NewSiteForm.jsx
  - Updated to use correct params ```params.siteId``` instead of ```parentId```

### frontend/src/components/SiteFlexbox.jsx
  - Updated the link to correct url when user clicks on the site name

### frontend/src/pages/GrowerFlowerPage.jsx
  - Updated to use params ```params.siteId``` if used on urls that start with _/grower/:siteId_
  - Updated to fetch correct flowers
      - All user flowers with ```getUserFlowers``` if url does not have :siteId
      - All flowers related to the site with ```getFlowersBySite``` if url has :siteId
  - Includes the id of the site if the url has :siteId
      - reason: requested to include for easier future development

### frontend/src/pages/GrowerHomePage.jsx
  - Updated the home page for growers. Shows either the text 
      - ”Home”, if it is accessed through _/grower_, or
      - ”{site.name}: {site.id}”, if accessed through _/grower/:siteId_
  - Distinction for those two cases was made for future adaptability, and the site ID is shown for easier development.

### frontend/src/translations.json
  - Added translations for menu.sites and title.site

## Notes:
- Go back button in ```<GrowerSitesPage />``` does no longer work as intended in all cases, but it was left as it is since there were talks of removing it completely in the future anyway.

Eoyie commented:
Great job!

**General comments:**

App.jsx is definitely improved, being less cluttered.  I do wonder if the NavigationBar.jsx and ProtectedRoute.jsx should be in some other folder, than the base frontend/src location. I would maybe even consider the navigation to be a component, but I'm not so sure about the protected route one. I'm definitely just nitpicking here as I'm not that capable of giving any proper suggestions on the code itself.

For some reason, I wasn't really expecting you to update the navigation in the grower layout in this branch and just test the routes manually. Still it's nice that this issue of the Outlet working properly was solved as well. I don't imagine the merge conflicts between this and the navigation branch should be that difficult either way.

**Other ideas:**

- Moving to a subsite I think should direct to you that site's home page and not that subsites sites page.
- I think there should be more general routing for incorrect paths. By this I mean that right now if while logged in the user tried to get to /login they will redirected to the home page, but if they write any other incorrect path such as /site they're not brought to the homepage or shown any "incorrect path" message. This should probably be made into a separate issue?

amandahamynen commented:
Thank you for your review! I moved NavigationBar.jsx and ProtectedRoute.jsx into /components folder like you suggested. 

> For some reason, I wasn't really expecting you to update the navigation in the grower layout in this branch and just test the routes manually. Still it's nice that this issue of the Outlet working properly was solved as well. I don't imagine the merge conflicts between this and the navigation branch should be that difficult either way.

Since I needed to update the navigation to ensure it used the correct parameters, I went ahead and made the updates to the navigation. I wanted the app to be as polished as possible since these changes are going to the staging server before the UI update, so I thought it was worth making these small updates.

I also ended up making a small update to GrowerFlowerPage.jsx since i forgot to add ’Site: ’ in front of the ID (it was perhaps a bit unclear before that was was meant by the ID that was shown on that page)

> Moving to a subsite I think should direct to you that site's home page and not that subsites sites page.

I fixed this, thanks for pointing this out!

> I think there should be more general routing for incorrect paths. By this I mean that right now if while logged in the user tried to get to /login they will redirected to the home page, but if they write any other incorrect path such as /site they're not brought to the homepage or shown any "incorrect path" message. This should probably be made into a separate issue?

I agree, and i think this should be done in a seperate issue in the future. I’d imagine it to require atleast making a new component for ”NotFound” page, and then also adding routes for that, and i think it goes beyond the scope of this task.

Eoyie commented:
Thank you for the fixes! I checked out the made changes and the new sites to subsite homepage seems to be working just fine as well as /frontend/src/ being less cluttered now.

In my opinion this is now ready to be merged.
